src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/HeaderValueProvider.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Internal
+{
+    public class HeaderValueProvider : IValueProvider
+    {
+        private readonly CultureInfo _culture;
+        private readonly string _headerFieldName;
+        private readonly IHeaderDictionary _headers;
+
+        public HeaderValueProvider(
+            IHeaderDictionary headers,
+            CultureInfo culture,
+            string headerFieldName)
+        {
+            if (headers == null)
+            {
+                throw new ArgumentNullException(nameof(headers));
+            }
+
+            if (string.IsNullOrEmpty(headerFieldName))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(headerFieldName));
+            }
+
+            _headers = headers;
+            _culture = culture;
+            _headerFieldName = headerFieldName;
+        }
+
+        public bool UseCommaSeparatedValues { get; set; }
+
+        /// <inheritdoc />
+        public bool ContainsPrefix(string prefix)
+        {
+            // In 2.0 version, HeaderModelBinder didn't consider prefix at all and doing now would break existing
+            // users, so ignore the key value totally and only rely on the field name
+            return _headers.ContainsKey(_headerFieldName);
```

Comment:
We should be setting up the model binding context such that `key` is going to be header name, and then checking that it's the appropriate value.

This implementation doesn't really conform to the expectations of the `IValueProvider` interface.

Reply:
Yes, you are right and that's how I had it in the beginning, but I changed it because the [current ](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs#L96)HeaderModelBinder sets the key to be the `ModelName` instead of the header name and changing it might break existing users. So the idea, as you can imagine, is to send the inner binders the model name (so that they can set the model state key as before), but make the header value provider just ignore the full model path.